File: turbo/packages/api-contracts/src/contracts/oom-evidence.ts
import { z } from "zod";

// Same bounded metadata contract as guest-contracts::oom_evidence. Null means
// unavailable, including kernels that do not expose a particular field.
const counter = z.number().int().nonnegative().nullable();
const status = z.enum([
  "available",
  "missing",
  "partial",
  "unavailable",
  "denied",
  "truncated",
  "overwritten",
  "uncorrelated",
  "recreated",
]);
const cgroup = z
  .string()
  .max(256)
  .regex(
    /^\/vm0-exec\/exec-[0-9-]+\/workload(?:\/(?:runtime|tools(?:\/tool-[0-9-]+)?))?$/,
  );
const limit = z
  .string()
  .max(20)
  .regex(/^(?:max|[0-9]+)$/)
  .nullable();
const events = z
  .object({
    high: counter,
    max: counter,
    oom: counter,
    oom_kill: counter,
    oom_group_kill: counter,
  })
  .strict();
const memorySnapshotSchema = z
  .object({
    role: z.enum(["workload", "runtime", "tools"]),
    cgroup,
    inode: counter,
    status,
    current: counter,
    peak: counter,
    limit,
    initial_limit: limit,
    anon: counter,
    file: counter,
    kernel: counter,
    baseline: events,
    events,
    delta: events,
    local_baseline: events,
    local_events: events,
    local_delta: events,
  })
  .strict();
const groups = z.tuple([
  memorySnapshotSchema,
  memorySnapshotSchema,
  memorySnapshotSchema,
]);
const kernelEvent = z
  .object({
    source: z.literal("guest"),
    sequence: z.number().int().nonnegative(),
    boottime_us: z.number().int().nonnegative(),
    constraint: z.enum([
      "CONSTRAINT_NONE",
      "CONSTRAINT_MEMCG",
      "CONSTRAINT_CPUSET",
      "CONSTRAINT_MEMORY_POLICY",
    ]),
    oom_cgroup: z
      .union([
        cgroup,
        z.string().regex(/^(?:\/|\/vm0-exec(?:\/exec-[0-9-]+)?)$/),
      ])
      .nullable(),
    victim_pid: z.number().int().positive().max(0xffff_ffff),
    victim_comm: z
      .string()
      .min(1)
      .max(16)
      .refine((value) => {
        return ![...value].some((character) => {
          const code = character.codePointAt(0);
          return code !== undefined && (code < 32 || code === 127);
        });
      }),
    task_cgroup: z.union([
      cgroup,
      z
        .string()
        .max(256)
        .regex(/^\/vm0-exec\/exec-[0-9-]+\/control$/),
    ]),
  })
  .strict();
const incident = z
  .object({
    id: z
      .string()
      .max(40)
      .regex(/^[0-9a-f-]{36}:[1-4]$/),
    captured_at: z.iso.datetime(),
    reason: z.enum(["sample", "cli_error", "cleanup"]),
    after_observation: z.literal(true),
    before_cleanup: z.literal(true),
    kernel_status: status,
    kernel_events: z.array(kernelEvent).max(4),
    groups,
  })
  .strict();

export const oomEvidenceSchema = z
  .object({
    operation_id: z.uuid(),
    guest_boot_id: z.uuid().nullable(),
    started_boottime_us: z.number().int().nonnegative(),
    sampled_at: z.iso.datetime(),
    kernel_cursor: counter,
    kernel_status: status,
    groups,
    incidents: z.array(incident).max(4),
    dropped_incidents: z.number().int().nonnegative(),
  })
  .strict()
  .superRefine((evidence, context) => {
    const workload = evidence.groups[0].cgroup;
    const operationRoot = workload.slice(0, -"/workload".length);
    const validGroups = (value: z.infer<typeof groups>) => {
      return (
        value[0].role === "workload" &&
        value[0].cgroup === workload &&
        value[1].role === "runtime" &&
        value[1].cgroup === `${workload}/runtime` &&
        value[2].role === "tools" &&
        value[2].cgroup === `${workload}/tools`
      );
    };
    if (
      !validGroups(evidence.groups) ||
      evidence.incidents.some((item) => {
        return (
          !item.id.startsWith(`${evidence.operation_id}:`) ||
          !validGroups(item.groups) ||
          item.kernel_events.some((event) => {
            return (
              event.boottime_us < evidence.started_boottime_us ||
              (event.oom_cgroup !== null &&
                event.oom_cgroup !== "/" &&
                event.oom_cgroup !== "/vm0-exec" &&
                event.oom_cgroup !== operationRoot &&
                event.oom_cgroup !== workload &&
                !event.oom_cgroup.startsWith(`${workload}/`)) ||
              !(
                event.task_cgroup === workload ||
                event.task_cgroup ===
                  `${workload.slice(0, -"/workload".length)}/control` ||
                event.task_cgroup.startsWith(`${workload}/`)
              )
            );
          })
        );
      })
    ) {
      context.addIssue({
        code: "custom",
        message: "Evidence must belong to one containment lifecycle",
      });
    }
  });

File: crates/guest-contracts/src/oom_evidence.rs
//! Bounded, metadata-only evidence for one Guest containment lifecycle.

use serde::{Deserialize, Serialize};
use std::io::{self, Read, Write};
use std::os::unix::net::UnixStream;
use std::time::{Duration, Instant};

/// Maximum encoded evidence response, including all retained incidents.
pub const MAX_EVIDENCE_BYTES: usize = 48 * 1024;
/// Maximum retained incidents in an operation (additional incidents are counted).
pub const MAX_INCIDENTS: usize = 4;
/// Maximum kernel records retained in one incident.
pub const MAX_KERNEL_EVENTS: usize = 4;
/// Absolute budget for one local evidence exchange.
pub const EVIDENCE_IO_TIMEOUT: Duration = Duration::from_millis(200);
/// Prefix on the existing trusted Guest-control terminal diagnostic transport.
pub const EVIDENCE_PREFIX: &str = "OKOU_OOM_EVIDENCE_V1 ";

/// Why a bounded capture was requested. An error is not evidence of OOM.
#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CaptureReason {
    /// Existing five-second metrics tick.
    Sample,
    /// Abnormal CLI completion.
    CliError,
    /// Containment is about to signal/remove its descendants.
    Cleanup,
}

/// Availability is separate from measured zero.
#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EvidenceStatus {
    /// All requested source data was available.
    Available,
    /// The reader succeeded but no attributable kernel record was present.
    Missing,
    /// At least one optional source field was absent.
    Partial,
    /// Source does not exist or cannot be read.
    Unavailable,
    /// Source access was denied.
    Denied,
    /// A byte, count, or time budget was exhausted.
    Truncated,
    /// Kernel records were overwritten before the cursor could read them.
    Overwritten,
    /// A source record could not be attributed to this operation.
    Uncorrelated,
    /// The operation-owned cgroup inode changed.
    Recreated,
}

/// Allowlisted memory event counters. Missing keys remain null.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct MemoryEvents {
    /// memory.high events.
    pub high: Option<u64>,
    /// memory.max events, not the configured byte limit.
    pub max: Option<u64>,
    /// OOM allocation failures.
    pub oom: Option<u64>,
    /// OOM-killed tasks.
    pub oom_kill: Option<u64>,
    /// OOM-killed groups, when supported.
    pub oom_group_kill: Option<u64>,
}

impl MemoryEvents {
    /// Subtract only known, non-regressing counters.
    pub fn delta(&self, baseline: &Self) -> Self {
        fn subtract(value: Option<u64>, baseline: Option<u64>) -> Option<u64> {
            value?.checked_sub(baseline?)
        }
        Self {
            high: subtract(self.high, baseline.high),
            max: subtract(self.max, baseline.max),
            oom: subtract(self.oom, baseline.oom),
            oom_kill: subtract(self.oom_kill, baseline.oom_kill),
            oom_group_kill: subtract(self.oom_group_kill, baseline.oom_group_kill),
        }
    }

    /// Whether a known OOM counter increased.
    pub fn has_oom(&self) -> bool {
        [self.oom, self.oom_kill, self.oom_group_kill]
            .into_iter()
            .flatten()
            .any(|value| value > 0)
    }
}

/// A fixed cgroup sample; never an inventory of processes or tool content.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct MemorySnapshot {
    /// Fixed role: workload, runtime, or tools.
    pub role: String,
    /// Trusted operation-owned cgroup path.
    pub cgroup: String,
    /// Kernel filesystem identity captured at initialization.
    pub inode: Option<u64>,
    /// Completeness of the source reads.
    pub status: EvidenceStatus,
    /// Current charged bytes, measured after an observed incident.
    pub current: Option<u64>,
    /// Kernel-maintained peak bytes when supported.
    pub peak: Option<u64>,
    /// Actual memory.max readback: decimal bytes or "max".
    pub limit: Option<String>,
    /// Initial limit readback.
    pub initial_limit: Option<String>,
    /// Anonymous charged bytes.
    pub anon: Option<u64>,
    /// File-backed charged bytes.
    pub file: Option<u64>,
    /// Kernel charged bytes when supported.
    pub kernel: Option<u64>,
    /// Initial hierarchical event counters.
    pub baseline: MemoryEvents,
    /// Current hierarchical event counters.
    pub events: MemoryEvents,
    /// Current minus initial hierarchical counters.
    pub delta: MemoryEvents,
    /// Initial local event counters.
    pub local_baseline: MemoryEvents,
    /// Current local event counters.
    pub local_events: MemoryEvents,
    /// Current minus initial local counters.
    pub local_delta: MemoryEvents,
}

/// Source-backed fields from a kernel-facility /dev/kmsg OOM record.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct KernelOomEvent {
    /// Always "guest"; Host records cannot enter this reader.
    pub source: String,
    /// Kernel ring sequence, separate from ingestion order.
    pub sequence: u64,
    /// Kernel record time in microseconds since boot.
    pub boottime_us: u64,
    /// Allowlisted kernel constraint name.
    pub constraint: String,
    /// Trigger cgroup only when it is a canonical ancestor of this workload.
    pub oom_cgroup: Option<String>,
    /// Victim PID reported by the kernel; no /proc lookup is required.
    pub victim_pid: u32,
    /// Bounded task comm from the kernel, never argv.
    pub victim_comm: String,
    /// Canonical operation-owned victim task cgroup from the same record.
    pub task_cgroup: String,
}

/// One deduplicated incident. A candidate can have no proof of OOM.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct OomIncident {
    /// Stable operation UUID plus incident ordinal.
    pub id: String,
    /// Capture time, distinct from source event and ingestion times.
    pub captured_at: String,
    /// Capture trigger; it does not alter execution outcome.
    pub reason: CaptureReason,
    /// Snapshots are taken after observation, never claimed as pre-kill usage.
    pub after_observation: bool,
    /// Snapshot precedes containment signaling/removal.
    pub before_cleanup: bool,
    /// Kernel reader completeness, independent of counter evidence.
    pub kernel_status: EvidenceStatus,
    /// Allowlisted kernel records, bounded by MAX_KERNEL_EVENTS.
    pub kernel_events: Vec<KernelOomEvent>,
    /// Exactly workload, runtime, and aggregate tools.
    pub groups: [MemorySnapshot; 3],
}

/// Operation identity and bounded retained evidence carried by both transports.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct OomEvidence {
    /// Fresh UUID for this containment creation, independent of sandbox/PID reuse.
    pub operation_id: String,
    /// Guest boot UUID; null if inaccessible.
    pub guest_boot_id: Option<String>,
    /// Monotonic initialization boundary for rejecting old kernel records.
    pub started_boottime_us: u64,
    /// Source sample time.
    pub sampled_at: String,
    /// Last consumed kernel sequence, absent before the first record.
    pub kernel_cursor: Option<u64>,
    /// Availability of the bounded kernel reader at this sample.
    pub kernel_status: EvidenceStatus,
    /// Current source sample, including identity and baseline.
    pub groups: [MemorySnapshot; 3],
    /// Bounded retained incidents.
    pub incidents: Vec<OomIncident>,
    /// Additional incidents beyond the retention cap.
    pub dropped_incidents: u64,
}

/// Exchange a bounded response on the authenticated placement connection.
pub fn read_evidence(stream: &UnixStream) -> io::Result<OomEvidence> {
    let deadline = Instant::now() + EVIDENCE_IO_TIMEOUT;
    let mut header = [0; 4];
    read_exact_until(stream, &mut header, deadline)?;
    let length = u32::from_be_bytes(header) as usize;
    if length > MAX_EVIDENCE_BYTES {
        return Err(io::Error::other("evidence response exceeds byte limit"));
    }
    let mut bytes = vec![0; length];
    read_exact_until(stream, &mut bytes, deadline)?;
    let evidence: OomEvidence = serde_json::from_slice(&bytes)?;
    if evidence.incidents.len() > MAX_INCIDENTS
        || evidence
            .incidents
            .iter()
            .any(|incident| incident.kernel_events.len() > MAX_KERNEL_EVENTS)
    {
        return Err(io::Error::other("evidence response exceeds count limit"));
    }
    Ok(evidence)
}

fn read_exact_until(
    mut stream: &UnixStream,
    mut bytes: &mut [u8],
    deadline: Instant,
) -> io::Result<()> {
    while !bytes.is_empty() {
        let remaining = deadline
            .checked_duration_since(Instant::now())
            .ok_or_else(|| io::Error::from(io::ErrorKind::TimedOut))?;
        stream.set_read_timeout(Some(remaining))?;
        let count = stream.read(bytes)?;
        if count == 0 {
            return Err(io::ErrorKind::UnexpectedEof.into());
        }
        bytes = bytes
            .get_mut(count..)
            .ok_or_else(|| io::Error::other("invalid evidence read length"))?;
    }
    Ok(())
}

/// Write one bounded response, failing closed on oversize evidence.
pub fn write_evidence(stream: &UnixStream, evidence: &OomEvidence) -> io::Result<()> {
    let bytes = serde_json::to_vec(evidence)?;
    if bytes.len() > MAX_EVIDENCE_BYTES {
        return Err(io::Error::other("evidence exceeds byte limit"));
    }
    let deadline = Instant::now() + EVIDENCE_IO_TIMEOUT;
    write_all_until(stream, &(bytes.len() as u32).to_be_bytes(), deadline)?;
    write_all_until(stream, &bytes, deadline)
}

fn write_all_until(mut stream: &UnixStream, mut bytes: &[u8], deadline: Instant) -> io::Result<()> {
    while !bytes.is_empty() {
        let remaining = deadline
            .checked_duration_since(Instant::now())
            .ok_or_else(|| io::Error::from(io::ErrorKind::TimedOut))?;
        stream.set_write_timeout(Some(remaining))?;
        let count = stream.write(bytes)?;
        if count == 0 {
            return Err(io::ErrorKind::WriteZero.into());
        }
        bytes = bytes
            .get(count..)
            .ok_or_else(|| io::Error::other("invalid evidence write length"))?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    const FIXTURE: &str = include_str!("../tests/fixtures/oom-evidence-v1.json");

    #[test]
    fn serialized_evidence_matches_the_cross_language_fixture() {
        let evidence: OomEvidence = serde_json::from_str(FIXTURE).unwrap();
        assert_eq!(
            serde_json::to_value(evidence).unwrap(),
            serde_json::from_str::<serde_json::Value>(FIXTURE).unwrap()
        );
    }

    #[test]
    fn evidence_exchange_preserves_identity_and_rejects_oversized_frames() {
        let (client, server) = UnixStream::pair().unwrap();
        let evidence: OomEvidence = serde_json::from_str(FIXTURE).unwrap();
        write_evidence(&server, &evidence).unwrap();
        assert_eq!(read_evidence(&client).unwrap(), evidence);
        (&server)
            .write_all(&((MAX_EVIDENCE_BYTES + 1) as u32).to_be_bytes())
            .unwrap();
        assert!(read_evidence(&client).is_err());
    }

    #[test]
    fn unavailable_evidence_peer_has_a_bounded_deadline() {
        let (client, _server) = UnixStream::pair().unwrap();
        let start = Instant::now();
        assert!(read_evidence(&client).is_err());
        assert!(start.elapsed() < Duration::from_secs(2));
    }

    #[test]
    fn evidence_exchange_rejects_count_overflow_and_bounds_a_stalled_writer() {
        let (client, server) = UnixStream::pair().unwrap();
        let mut evidence: OomEvidence = serde_json::from_str(FIXTURE).unwrap();
        evidence.incidents = vec![evidence.incidents[0].clone(); MAX_INCIDENTS + 1];
        write_evidence(&server, &evidence).unwrap();
        assert!(read_evidence(&client).is_err());
        evidence.incidents.truncate(1);
        evidence.incidents[0].kernel_events =
            vec![evidence.incidents[0].kernel_events[0].clone(); MAX_KERNEL_EVENTS + 1];
        write_evidence(&server, &evidence).unwrap();
        assert!(read_evidence(&client).is_err());
        evidence.operation_id = "x".repeat(MAX_EVIDENCE_BYTES);
        assert!(write_evidence(&server, &evidence).is_err());
        let start = Instant::now();
        assert!(
            write_all_until(&server, &vec![0; 1024 * 1024], start + EVIDENCE_IO_TIMEOUT).is_err()
        );
        assert!(start.elapsed() < Duration::from_secs(2));
    }

    #[test]
    fn counter_regression_and_missing_baselines_are_unavailable() {
        let evidence: OomEvidence = serde_json::from_str(FIXTURE).unwrap();
        let current = &evidence.groups[0].events;
        assert_eq!(MemoryEvents::default().delta(current).oom_kill, None);
        assert_eq!(current.delta(&MemoryEvents::default()).oom_kill, None);
        assert!(!current.delta(current).has_oom());
    }
}
